File: typescript/packages/server/src/ssh/sftp.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { constants as fsConstants } from 'node:fs'
import { constants as osConstants } from 'node:os'
import { posix } from 'node:path'
import { MountCore, classifyErrno, type FuseAttr } from '@struktoai/mirage-node'
import { EACCES, ENOENT, EROFS, errnoError } from '@struktoai/mirage-node/fuse/errors'
import type { Attributes, FileEntry, SFTPWrapper } from 'ssh2'
import type { WorkspaceEntry, WorkspaceRegistry } from '../registry.ts'
import { SFTPStatusError } from './errors.ts'
import { newSessionId, openSession } from './session.ts'

// SFTP v3 open flags and status codes, fixed by the protocol draft.
const OPEN = {
  READ: 0x01,
  WRITE: 0x02,
  APPEND: 0x04,
  CREAT: 0x08,
  TRUNC: 0x10,
  EXCL: 0x20,
} as const
export const STATUS = {
  OK: 0,
  EOF: 1,
  NO_SUCH_FILE: 2,
  PERMISSION_DENIED: 3,
  FAILURE: 4,
  OP_UNSUPPORTED: 8,
} as const

const S_IFMT = 0o170000
const S_IFDIR = 0o040000
const S_IFLNK = 0o120000
const MONTHS = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
const NO_ATTRS = {} as Attributes

interface OpenFile {
  kind: 'file'
  path: string
  fd: number
  appendAt: number | null
}

interface OpenDir {
  kind: 'dir'
  path: string
  done: boolean
}

type OpenHandle = OpenFile | OpenDir

function isDir(mode: number): boolean {
  return (mode & S_IFMT) === S_IFDIR
}

function seconds(date: Date): number {
  return Math.floor(date.getTime() / 1000)
}

/** SFTP attributes from a MountCore attr record. */
export function toAttrs(a: FuseAttr): Attributes {
  return {
    mode: a.mode,
    uid: a.uid,
    gid: a.gid,
    size: a.size,
    atime: seconds(a.atime),
    mtime: seconds(a.mtime),
  }
}

function permissions(mode: number): string {
  const kind = isDir(mode) ? 'd' : (mode & S_IFMT) === S_IFLNK ? 'l' : '-'
  let bits = ''
  for (let shift = 6; shift >= 0; shift -= 3) {
    const triplet = (mode >> shift) & 0o7
    bits += (triplet & 4 ? 'r' : '-') + (triplet & 2 ? 'w' : '-') + (triplet & 1 ? 'x' : '-')
  }
  return kind + bits
}

/** The `ls -l` line an SFTP v3 client prints for an entry. */
export function longname(name: string, a: Attributes): string {
  const when = new Date(a.mtime * 1000)
  const day = String(when.getDate()).padStart(2)
  const time = `${String(when.getHours()).padStart(2, '0')}:${String(when.getMinutes()).padStart(2, '0')}`
  const month = MONTHS[when.getMonth()] ?? ''
  return `${permissions(a.mode)}    1 ${String(a.uid).padEnd(8)} ${String(a.gid).padEnd(8)} ${String(a.size).padStart(8)} ${month} ${day} ${time} ${name}`
}

/** A client path as a workspace path: absolute, normalized, never above `/`. */
export function workspacePath(path: string): string {
  const normal = posix.normalize('/' + path.replace(/^\/+/, ''))
  return normal.length > 1 && normal.endsWith('/') ? normal.slice(0, -1) : normal
}

async function exists(core: MountCore, path: string): Promise<boolean> {
  try {
    await core.getattr(path)
    return true
  } catch (err) {
    if (classifyErrno(err) === ENOENT) return false
    throw err
  }
}

const DENIED = new Set([EACCES, EROFS, osConstants.errno.EPERM])

/** The SFTP status and message for a failure, through the shared errno table. */
function statusOf(err: unknown): [number, string] {
  if (err instanceof SFTPStatusError) return [err.status, err.message]
  const code = classifyErrno(err)
  const message = err instanceof Error ? err.message : String(err)
  if (code === ENOENT) return [STATUS.NO_SUCH_FILE, message]
  if (DENIED.has(code)) return [STATUS.PERMISSION_DENIED, message]
  return [STATUS.FAILURE, message]
}

// ssh2 ends an SFTP channel with no exit status, which OpenSSH's ssh
// reports as 255 and scp (in its default SFTP mode) as a failed copy.
// sshd's internal-sftp reports 0; this sends it the way ssh2's own
// Channel.exit does for a session channel.
interface SFTPChannelInternals {
  _protocol: { exitStatus(id: number, status: number): void }
  outgoing: { id: number; state: string }
}

function exitZero(sftp: SFTPWrapper): void {
  const chan = sftp as unknown as SFTPChannelInternals
  if (chan.outgoing.state === 'open') chan._protocol.exitStatus(chan.outgoing.id, 0)
}

/**
 * SFTP onto a workspace, through the MountCore FUSE uses.
 *
 * Every request lands on one MountCore bound to a session of its own, so
 * SFTP sees exactly the tree, modes and policies a shell in that session
 * sees. Requests are answered one at a time in arrival order: a client
 * pipelines writes and then closes, and the close must not overtake the
 * writes it settles. ssh2 refuses any request type left unhandled, so
 * nothing here can fall through to the host filesystem.
 */
class MirageSFTPServer {
  private readonly handles = new Map<string, OpenHandle>()
  private nextHandle = 0
  private readonly sessionId = newSessionId()
  private entry: WorkspaceEntry | null = null
  private core: MountCore | null = null
  private queue: Promise<void> = Promise.resolve()

  constructor(
    private readonly registry: WorkspaceRegistry,
    private readonly workspaceId: string,
    private readonly sftp: SFTPWrapper,
  ) {}

  attach(): void {
    const s = this.sftp
    s.on('REALPATH', (id, path) => {
      this.serve(id, () => this.realpath(id, path))
    })
    s.on('STAT', (id, path) => {
      this.serve(id, () => this.stat(id, path, true))
    })
    s.on('LSTAT', (id, path) => {
      this.serve(id, () => this.stat(id, path, false))
    })
    s.on('FSTAT', (id, handle) => {
      this.serve(id, () => this.fstat(id, handle))
    })
    s.on('SETSTAT', (id, path, attrs) => {
      this.serve(id, () => this.setSize(id, workspacePath(path), attrs))
    })
    s.on('FSETSTAT', (id, handle, attrs) => {
      this.serve(id, () => this.setSize(id, this.file(handle).path, attrs))
    })
    s.on('OPENDIR', (id, path) => {
      this.serve(id, () => this.opendir(id, path))
    })
    s.on('READDIR', (id, handle) => {
      this.serve(id, () => this.readdir(id, handle))
    })
    s.on('OPEN', (id, filename, flags) => {
      this.serve(id, () => this.open(id, filename, flags))
    })
    s.on('READ', (id, handle, offset, len) => {
      this.serve(id, () => this.read(id, handle, offset, len))
    })
    s.on('WRITE', (id, handle, offset, data) => {
      this.serve(id, () => this.write(id, handle, offset, data))
    })
    s.on('CLOSE', (id, handle) => {
      this.serve(id, () => this.close(id, handle))
    })
    s.on('REMOVE', (id, path) => {
      this.serve(id, () => this.simple(id, (core) => core.unlink(workspacePath(path))))
    })
    s.on('MKDIR', (id, path) => {
      this.serve(id, () => this.simple(id, (core) => core.mkdir(workspacePath(path))))
    })
    s.on('RMDIR', (id, path) => {
      this.serve(id, () => this.simple(id, (core) => core.rmdir(workspacePath(path))))
    })
    s.on('RENAME', (id, oldPath, newPath) => {
      this.serve(id, () => this.rename(id, workspacePath(oldPath), workspacePath(newPath)))
    })
    s.on('READLINK', (id, path) => {
      this.serve(id, () => this.readlink(id, path))
    })
    // ssh2 emits (link, target) whichever order the client sent them in
    // (it undoes OpenSSH's swap), as its README documents; @types/ssh2
    // names the two parameters the other way round.
    s.on('SYMLINK', (id, linkPath, targetPath) => {
      this.serve(id, () =>
        this.simple(id, (core) => core.symlink(targetPath, workspacePath(linkPath))),
      )
    })
    s.on('EXTENDED', (id) => {
      this.sftp.status(id, STATUS.OP_UNSUPPORTED, 'extension not supported')
    })
    s.on('end', () => {
      this.queue = this.queue.then(() => this.exit())
    })
  }

  private serve(id: number, task: () => Promise<void>): void {
    this.queue = this.queue.then(async () => {
      try {
        await task()
      } catch (err) {
        const [status, message] = statusOf(err)
        this.sftp.status(id, status, message)
      }
    })
  }

  private async mount(): Promise<MountCore> {
    if (this.core !== null) return this.core
    if (!this.registry.has(this.workspaceId)) {
      throw new SFTPStatusError(STATUS.NO_SUCH_FILE, `no such workspace: ${this.workspaceId}`)
    }
    const entry = this.registry.get(this.workspaceId)
    const ws = entry.runner.ws
    await openSession(ws, this.sessionId)
    this.entry = entry
    this.core = new MountCore(ws.vfs, { session: ws.getSession(this.sessionId) })
    return this.core
  }

  private addHandle(handle: OpenHandle): Buffer {
    const id = Buffer.alloc(4)
    id.writeUInt32BE(this.nextHandle)
    this.nextHandle = (this.nextHandle + 1) >>> 0
    this.handles.set(id.toString('hex'), handle)
    return id
  }

  private handle(buf: Buffer): OpenHandle {
    const handle = this.handles.get(buf.toString('hex'))
    if (handle === undefined) throw new SFTPStatusError(STATUS.FAILURE, 'invalid handle')
    return handle
  }

  private file(buf: Buffer): OpenFile {
    const handle = this.handle(buf)
    if (handle.kind !== 'file') throw new SFTPStatusError(STATUS.FAILURE, 'not a file handle')
    return handle
  }

  private async simple(id: number, op: (core: MountCore) => Promise<void>): Promise<void> {
    await op(await this.mount())
    this.sftp.status(id, STATUS.OK)
  }

  private async realpath(id: number, path: string): Promise<void> {
    await this.mount()
    const p = workspacePath(path)
    this.sftp.name(id, [{ filename: p, longname: p, attrs: NO_ATTRS }])
  }

  private async stat(id: number, path: string, follow: boolean): Promise<void> {
    const core = await this.mount()
    const p = workspacePath(path)
    this.sftp.attrs(id, toAttrs(await core.getattr(follow ? core.identity(p) : p)))
  }

  private async fstat(id: number, buf: Buffer): Promise<void> {
    const core = await this.mount()
    const f = this.file(buf)
    this.sftp.attrs(id, toAttrs(await core.fgetattr(f.path, f.fd)))
  }

  // A size truncates; any other attribute is accepted once the path is
  // known to exist, as the FUSE adapter treats chmod, chown and utimens.
  private async setSize(id: number, path: string, attrs: Partial<Attributes>): Promise<void> {
    const core = await this.mount()
    if (typeof attrs.size === 'number') await core.truncate(path, attrs.size)
    else await core.getattr(path)
    this.sftp.status(id, STATUS.OK)
  }

  private async opendir(id: number, path: string): Promise<void> {
    const core = await this.mount()
    const p = workspacePath(path)
    if (!isDir((await core.getattr(p)).mode)) throw errnoError('ENOTDIR', p)
    this.sftp.handle(id, this.addHandle({ kind: 'dir', path: p, done: false }))
  }

  private async readdir(id: number, buf: Buffer): Promise<void> {
    const core = await this.mount()
    const dir = this.handle(buf)
    if (dir.kind !== 'dir') throw new SFTPStatusError(STATUS.FAILURE, 'not a directory handle')
    if (dir.done) {
      this.sftp.status(id, STATUS.EOF)
      return
    }
    dir.done = true
    const entries: FileEntry[] = []
    for (const name of await core.readdir(dir.path)) {
      const child =
        name === '.'
          ? dir.path
          : name === '..'
            ? posix.dirname(dir.path)
            : posix.join(dir.path, name)
      let attr: FuseAttr
      try {
        attr = await core.getattr(child)
      } catch (err) {
        // Vanished between the listing and its stat, as ls leaves out a
        // file deleted mid-listing; anything else is a real failure.
        if (classifyErrno(err) !== ENOENT) throw err
        continue
      }
      const attrs = toAttrs(attr)
      entries.push({ filename: name, longname: longname(name, attrs), attrs })
    }
    this.sftp.name(id, entries)
  }

  private async open(id: number, filename: string, flags: number): Promise<void> {
    const core = await this.mount()
    const path = workspacePath(filename)
    const found = await exists(core, path)
    if (found && (flags & OPEN.CREAT) !== 0 && (flags & OPEN.EXCL) !== 0) {
      throw errnoError('EEXIST', path)
    }
    let fd: number
    if (!found) {
      if ((flags & OPEN.CREAT) === 0) throw errnoError('ENOENT', path)
      fd = await core.create(path)
    } else {
      if (isDir((await core.getattr(path)).mode)) throw errnoError('EISDIR', path)
      fd = await core.open(path, (flags & OPEN.TRUNC) !== 0 ? fsConstants.O_TRUNC : 0)
    }
    const appendAt = (flags & OPEN.APPEND) !== 0 ? (await core.fgetattr(path, fd)).size : null
    this.sftp.handle(id, this.addHandle({ kind: 'file', path, fd, appendAt }))
  }

  private async read(id: number, buf: Buffer, offset: number, len: number): Promise<void> {
    const core = await this.mount()
    const f = this.file(buf)
    const data = await core.read(f.path, f.fd, offset, len)
    if (data.byteLength === 0) {
      this.sftp.status(id, STATUS.EOF)
      return
    }
    this.sftp.data(id, Buffer.from(data.buffer, data.byteOffset, data.byteLength))
  }

  private async write(id: number, buf: Buffer, offset: number, data: Buffer): Promise<void> {
    const core = await this.mount()
    const f = this.file(buf)
    let at = offset
    if (f.appendAt !== null) {
      at = f.appendAt
      f.appendAt += data.byteLength
    }
    // A copy: the handle holds the bytes until close, and ssh2 reuses the
    // packet buffer they arrived in.
    await core.write(f.path, f.fd, new Uint8Array(data), at)
    this.sftp.status(id, STATUS.OK)
  }

  private async close(id: number, buf: Buffer): Promise<void> {
    const core = await this.mount()
    const handle = this.handle(buf)
    this.handles.delete(buf.toString('hex'))
    if (handle.kind === 'file') await core.release(handle.fd)
    this.sftp.status(id, STATUS.OK)
  }

  // SFTP v3's rename refuses to replace an existing target.
  private async rename(id: number, oldPath: string, newPath: string): Promise<void> {
    const core = await this.mount()
    if (await exists(core, newPath)) throw errnoError('EEXIST', newPath)
    await core.rename(oldPath, newPath)
    this.sftp.status(id, STATUS.OK)
  }

  private async readlink(id: number, path: string): Promise<void> {
    const core = await this.mount()
    const target = core.readlink(workspacePath(path))
    this.sftp.name(id, [{ filename: target, longname: target, attrs: NO_ATTRS }])
  }

  private async exit(): Promise<void> {
    const core = this.core
    const entry = this.entry
    try {
      if (core !== null) {
        for (const handle of this.handles.values()) {
          if (handle.kind === 'file') await core.release(handle.fd)
        }
      }
      this.handles.clear()
      if (entry !== null && this.registry.has(entry.id) && this.registry.get(entry.id) === entry) {
        await entry.runner.ws.closeSession(this.sessionId)
      }
    } finally {
      exitZero(this.sftp)
      this.sftp.end()
    }
  }
}

/** Serve SFTP (and so modern scp) for one channel onto its workspace. */
export function serveSFTP(
  registry: WorkspaceRegistry,
  workspaceId: string,
  sftp: SFTPWrapper,
): void {
  new MirageSFTPServer(registry, workspaceId, sftp).attach()
}

File: python/mirage/server/ssh/sftp.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
import errno
import logging
import os
import posixpath
import stat
from collections.abc import AsyncIterator, Callable
from dataclasses import dataclass
from typing import Any, TypeVar

import asyncssh
from asyncssh.constants import (FILEXFER_TYPE_DIRECTORY, FILEXFER_TYPE_REGULAR,
                                FILEXFER_TYPE_SYMLINK, FILEXFER_TYPE_UNKNOWN,
                                FXF_APPEND, FXF_CREAT, FXF_EXCL, FXF_TRUNC)

from mirage.fuse.core import MountCore
from mirage.fuse.errors import classify_error
from mirage.server.registry import WorkspaceEntry, WorkspaceRegistry
from mirage.server.ssh.session import new_session_id, open_session
from mirage.server.ssh.stream import ENCODING, ERRORS
from mirage.utils.errors import NoMountError

logger = logging.getLogger(__name__)

T = TypeVar("T")

NS_PER_SECOND = 1_000_000_000


@dataclass(slots=True)
class OpenFile:
    """An SFTP file handle: the MountCore handle behind it.

    Args:
        path (str): the workspace path it was opened on.
        fh (int): the MountCore handle id.
        append_at (int | None): where the next write lands when the file
            was opened for append, None otherwise.
    """

    path: str
    fh: int
    append_at: int | None = None


def opened(file_obj: Any) -> OpenFile:
    """Validate the opaque handle asyncssh passes to a file callback.

    ``Any`` at the library boundary keeps the overrides compatible with
    asyncssh. Only a validated ``OpenFile`` reaches the mount operations.

    Args:
        file_obj (Any): what asyncssh passed back.

    Returns:
        OpenFile: the handle.

    Raises:
        asyncssh.SFTPFailure: the value is not a handle this server made.
    """
    if not isinstance(file_obj, OpenFile):
        raise asyncssh.SFTPFailure("invalid handle")
    return file_obj


def filetype(mode: int) -> int:
    if stat.S_ISDIR(mode):
        return FILEXFER_TYPE_DIRECTORY
    if stat.S_ISLNK(mode):
        return FILEXFER_TYPE_SYMLINK
    if stat.S_ISREG(mode):
        return FILEXFER_TYPE_REGULAR
    return FILEXFER_TYPE_UNKNOWN


def to_attrs(st: dict[str, Any]) -> asyncssh.SFTPAttrs:
    """SFTP attributes from a MountCore ``st_*`` dict (times in ns).

    Args:
        st (dict[str, Any]): what ``MountCore.getattr`` returned.

    Returns:
        asyncssh.SFTPAttrs: the same facts in SFTP's shape.
    """
    mode = st["st_mode"]
    atime, atime_ns = divmod(st["st_atime"], NS_PER_SECOND)
    mtime, mtime_ns = divmod(st["st_mtime"], NS_PER_SECOND)
    return asyncssh.SFTPAttrs(type=filetype(mode),
                              size=st["st_size"],
                              uid=st["st_uid"],
                              gid=st["st_gid"],
                              permissions=mode,
                              atime=atime,
                              atime_ns=atime_ns,
                              mtime=mtime,
                              mtime_ns=mtime_ns,
                              nlink=st["st_nlink"])


def exists(core: MountCore, path: str) -> bool:
    try:
        core.getattr(path)
    except (FileNotFoundError, NotADirectoryError, NoMountError):
        return False
    return True


def listing(core: MountCore, path: str) -> list[tuple[str, dict[str, Any]]]:
    """A directory's entries with their attributes, in one pass.

    An entry that vanishes between the listing and its stat is left out,
    as ``ls`` leaves out a file deleted mid-listing.

    Args:
        core (MountCore): the mount core.
        path (str): the directory.

    Returns:
        list[tuple[str, dict[str, Any]]]: (name, ``st_*`` dict) pairs,
            ``.`` and ``..`` first.
    """
    entries = []
    for name in core.readdir(path):
        if name == ".":
            child = path
        elif name == "..":
            child = posixpath.dirname(path)
        else:
            child = posixpath.join(path, name)
        try:
            entries.append((name, core.getattr(child)))
        except (FileNotFoundError, NotADirectoryError) as exc:
            logger.debug("sftp: %s vanished while listing: %r", child, exc)
    return entries


def open_file(core: MountCore, path: str, pflags: int) -> OpenFile:
    """Open or create a file the way an SFTP ``open`` asks.

    Args:
        core (MountCore): the mount core.
        path (str): the file.
        pflags (int): SFTP v3 open flags.

    Returns:
        OpenFile: the new handle.

    Raises:
        FileExistsError: CREAT|EXCL on an existing path.
        FileNotFoundError: no such file and no CREAT.
        IsADirectoryError: the path is a directory.
    """
    found = exists(core, path)
    if found and pflags & FXF_CREAT and pflags & FXF_EXCL:
        raise FileExistsError(errno.EEXIST, os.strerror(errno.EEXIST), path)
    if not found:
        if not pflags & FXF_CREAT:
            raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT),
                                    path)
        fh = core.create(path)
    else:
        if stat.S_ISDIR(core.getattr(path)["st_mode"]):
            raise IsADirectoryError(errno.EISDIR, os.strerror(errno.EISDIR),
                                    path)
        fh = core.open(path, os.O_TRUNC if pflags & FXF_TRUNC else 0)
    append_at = None
    if pflags & FXF_APPEND:
        append_at = core.getattr(path, fh)["st_size"]
    return OpenFile(path, fh, append_at)


def set_size(core: MountCore, path: str, size: int | None) -> None:
    """Apply an SFTP setstat: a size truncates, anything else is accepted
    once the path is known to exist, as the FUSE adapter treats chmod,
    chown and utimens.

    Args:
        core (MountCore): the mount core.
        path (str): the path.
        size (int | None): the requested size, if any.
    """
    if size is not None:
        core.truncate(path, size)
    else:
        core.getattr(path)


def rename_new(core: MountCore, old: str, new: str) -> None:
    """SFTP v3 rename, which refuses to replace an existing target.

    Args:
        core (MountCore): the mount core.
        old (str): the current path.
        new (str): the new path.
    """
    if exists(core, new):
        raise FileExistsError(errno.EEXIST, os.strerror(errno.EEXIST), new)
    core.rename(old, new)


def vfs_attrs(core: MountCore) -> asyncssh.SFTPVFSAttrs:
    st = core.statfs()
    return asyncssh.SFTPVFSAttrs(bsize=st["f_bsize"],
                                 frsize=st["f_frsize"],
                                 blocks=st["f_blocks"],
                                 bfree=st["f_bfree"],
                                 bavail=st["f_bavail"],
                                 files=st["f_files"],
                                 ffree=st["f_ffree"],
                                 favail=st["f_favail"],
                                 fsid=0,
                                 flag=0,
                                 namemax=st["f_namemax"])


def as_os_error(err: Exception) -> OSError:
    """The errno asyncssh turns into an SFTP status, from the shared table.

    Args:
        err (Exception): what the core raised.

    Returns:
        OSError: carrying the classified errno.
    """
    code = classify_error(err)
    if code == errno.EIO and not isinstance(err, (OSError, ValueError)):
        logger.warning("sftp: unclassified error: %r", err)
    return OSError(code, os.strerror(code))


class MirageSFTPServer(asyncssh.SFTPServer):
    """SFTP (and scp) onto a workspace, through the MountCore FUSE uses.

    Every request lands on one MountCore bound to a session of its own,
    so SFTP sees exactly the tree, modes and policies a shell in that
    session sees. The core is synchronous (FUSE calls it from a single
    thread), so calls run one at a time in a worker thread, and the core
    runs each op on the workspace's own loop.

    asyncssh's base class serves the host's real filesystem from every
    method a subclass leaves alone, so this class overrides all of them
    and ``map_path`` refuses, keeping a future asyncssh method from
    reaching the host.

    Args:
        registry (WorkspaceRegistry): the daemon's workspaces; the SSH
            username names the one served.
        chan (asyncssh.SSHServerChannel): the SFTP channel.
    """

    def __init__(self, registry: WorkspaceRegistry,
                 chan: asyncssh.SSHServerChannel[bytes]) -> None:
        super().__init__(chan)
        self._registry = registry
        self._workspace_id: str = chan.get_extra_info("username")
        self._session_id = new_session_id()
        self._entry: WorkspaceEntry | None = None
        self._core: MountCore | None = None
        self._lock = asyncio.Lock()

    async def _mount(self) -> MountCore:
        if self._core is not None:
            return self._core
        if self._workspace_id not in self._registry:
            raise asyncssh.SFTPNoSuchFile(
                f"no such workspace: {self._workspace_id}")
        entry = self._registry.get(self._workspace_id)
        ws = entry.runner.ws
        await entry.runner.call(open_session(ws, self._session_id))
        self._entry = entry
        self._core = MountCore(ws.vfs,
                               session=ws.get_session(self._session_id),
                               loop=entry.runner.loop)
        return self._core

    async def _call(self, op: Callable[[MountCore], T]) -> T:
        async with self._lock:
            core = await self._mount()
            try:
                return await asyncio.to_thread(op, core)
            except Exception as err:
                raise as_os_error(err) from err

    def _path(self, path: bytes) -> str:
        text = path.decode(ENCODING, ERRORS)
        return posixpath.normpath("/" + text.lstrip("/"))

    def _encode(self, path: str) -> bytes:
        return path.encode(ENCODING, ERRORS)

    def map_path(self, path: bytes) -> bytes:
        raise asyncssh.SFTPOpUnsupported("host paths are not served")

    def reverse_map_path(self, path: bytes) -> bytes:
        raise asyncssh.SFTPOpUnsupported("host paths are not served")

    async def realpath(self, path: bytes) -> bytes:
        return self._encode(self._path(path))

    async def stat(self, path: bytes) -> asyncssh.SFTPAttrs:
        p = self._path(path)
        return to_attrs(
            await self._call(lambda core: core.getattr(core.identity(p))))

    async def lstat(self, path: bytes) -> asyncssh.SFTPAttrs:
        p = self._path(path)
        return to_attrs(await self._call(lambda core: core.getattr(p)))

    async def fstat(self, file_obj: Any) -> asyncssh.SFTPAttrs:
        f = opened(file_obj)
        return to_attrs(await
                        self._call(lambda core: core.getattr(f.path, f.fh)))

    async def setstat(self, path: bytes, attrs: asyncssh.SFTPAttrs) -> None:
        p = self._path(path)
        await self._call(lambda core: set_size(core, p, attrs.size))

    async def lsetstat(self, path: bytes, attrs: asyncssh.SFTPAttrs) -> None:
        await self.setstat(path, attrs)

    async def fsetstat(self, file_obj: Any, attrs: asyncssh.SFTPAttrs) -> None:
        f = opened(file_obj)

        def resize(core: MountCore) -> None:
            ctx = core.handles.get(f.fh)
            if ctx is None:
                raise asyncssh.SFTPFailure("invalid handle")
            set_size(core, ctx.path, attrs.size)

        await self._call(resize)

    async def scandir(self, path: bytes) -> AsyncIterator[asyncssh.SFTPName]:
        p = self._path(path)
        for name, st in await self._call(lambda core: listing(core, p)):
            yield asyncssh.SFTPName(self._encode(name), attrs=to_attrs(st))

    async def open(self, path: bytes, pflags: int,
                   attrs: asyncssh.SFTPAttrs) -> OpenFile:
        p = self._path(path)
        return await self._call(lambda core: open_file(core, p, pflags))

    async def open56(self, path: bytes, desired_access: int, flags: int,
                     attrs: asyncssh.SFTPAttrs) -> OpenFile:
        raise asyncssh.SFTPOpUnsupported("SFTP v5+ open is not supported")

    async def read(self, file_obj: Any, offset: int, size: int) -> bytes:
        f = opened(file_obj)
        return await self._call(
            lambda core: core.read(f.path, size, offset, f.fh))

    async def write(self, file_obj: Any, offset: int, data: bytes) -> int:
        f = opened(file_obj)
        if f.append_at is not None:
            offset = f.append_at
            f.append_at += len(data)
        return await self._call(
            lambda core: core.write(f.path, data, offset, f.fh))

    async def fsync(self, file_obj: Any) -> None:
        f = opened(file_obj)
        await self._call(lambda core: core.flush(f.path, f.fh))

    async def close(self, file_obj: Any) -> None:
        f = opened(file_obj)
        await self._call(lambda core: core.release(f.fh))

    async def remove(self, path: bytes) -> None:
        p = self._path(path)
        await self._call(lambda core: core.unlink(p))

    async def mkdir(self, path: bytes, attrs: asyncssh.SFTPAttrs) -> None:
        p = self._path(path)
        await self._call(lambda core: core.mkdir(p))

    async def rmdir(self, path: bytes) -> None:
        p = self._path(path)
        await self._call(lambda core: core.rmdir(p))

    async def rename(self, oldpath: bytes, newpath: bytes) -> None:
        old, new = self._path(oldpath), self._path(newpath)
        await self._call(lambda core: rename_new(core, old, new))

    async def posix_rename(self, oldpath: bytes, newpath: bytes) -> None:
        old, new = self._path(oldpath), self._path(newpath)
        await self._call(lambda core: core.rename(old, new))

    async def readlink(self, path: bytes) -> bytes:
        p = self._path(path)
        return self._encode(await self._call(lambda core: core.readlink(p)))

    async def symlink(self, oldpath: bytes, newpath: bytes) -> None:
        target = oldpath.decode(ENCODING, ERRORS)
        link = self._path(newpath)
        await self._call(lambda core: core.symlink(link, target))

    async def link(self, oldpath: bytes, newpath: bytes) -> None:
        raise asyncssh.SFTPOpUnsupported("hard links are not supported")

    async def lock(self, file_obj: Any, offset: int, length: int,
                   flags: int) -> None:
        raise asyncssh.SFTPOpUnsupported("byte-range locks are not supported")

    async def unlock(self, file_obj: Any, offset: int, length: int) -> None:
        raise asyncssh.SFTPOpUnsupported("byte-range locks are not supported")

    async def statvfs(self, path: bytes) -> asyncssh.SFTPVFSAttrs:
        return await self._call(vfs_attrs)

    async def fstatvfs(self, file_obj: Any) -> asyncssh.SFTPVFSAttrs:
        return await self._call(vfs_attrs)

    async def exit(self) -> None:
        entry = self._entry
        try:
            if entry is not None and entry.id in self._registry and (
                    self._registry.get(entry.id) is entry):
                await entry.runner.call(
                    entry.runner.ws.close_session(self._session_id))
        finally:
            # asyncssh ends an SFTP channel with no exit status, which
            # OpenSSH's ssh reports as 255 and scp (in its default SFTP
            # mode) as a failed copy. sshd's internal-sftp reports 0, so
            # this does too; after scp's own handler has already exited
            # the channel, it is a no-op.
            self.channel.exit(0)
